sql/catalyst/src/main/antlr4/org/apache/spark/sql/catalyst/parser/SqlBaseParser.g4

```diff
@@ -618,6 +618,46 @@ pivotValue
     : expression (AS? identifier)?
     ;
 
+unpivotClause
+    : UNPIVOT nullOperator=unpivotNullClause? LEFT_PAREN
+        operator=unpivotOperator
+      RIGHT_PAREN (AS? identifier)?
+    ;
+
+unpivotNullClause
+    : (INCLUDE | EXCLUDE) NULLS
+    ;
+
+unpivotOperator
+    : (unpivotSingleValueColumnClause | unpivotMultiValueColumnClause)
+    ;
+
+unpivotSingleValueColumnClause
+    : unpivotValueColumn FOR unpivotNameColumn IN LEFT_PAREN unpivotColumns+=unpivotColumn (COMMA unpivotColumns+=unpivotColumn)* RIGHT_PAREN
+    ;
+
+unpivotMultiValueColumnClause
+    : LEFT_PAREN unpivotValueColumns+=unpivotValueColumn (COMMA unpivotValueColumns+=unpivotValueColumn)* RIGHT_PAREN
+      FOR unpivotNameColumn
+      IN LEFT_PAREN unpivotColumnSets+=unpivotColumnSet (COMMA unpivotColumnSets+=unpivotColumnSet)* RIGHT_PAREN
+    ;
+
+unpivotColumnSet
+    : LEFT_PAREN unpivotColumns+=unpivotColumn (COMMA unpivotColumns+=unpivotColumn)* RIGHT_PAREN (AS? identifier)?
```

Comment:
do we allow `IN ((col1 AS a, col2 AS b) AS c)`?

Reply:
Yes, `unpivotColumn` itself is defined as `expression (AS? identifier)?`, so each individual column can have an alias, and the entire set can. However, with an alias for the set, the alias of the column is hidden. But the syntax is valid.

Reply:
is it also valid in other systems?

Reply:
It looks like other systems only allow for identifiers available in the `FROM`, not expressions as our unpivot supports. In those systems you would have to move the expressions into a subquery in `FROM` and reference the named expression.

Reply:
If this is a spark-specific syntax, shall we make it simple and don't allow per-expression alias here?

Reply:
ping @EnricoMi

Reply:
Saying `unpivotColumn` should be `namedExpression` instead of `expression (AS? identifier)?`?

Reply:
just `expression`? people can still use `col AS alias` but we just treat it as a normal expression.

Reply:
`expression` does not allow for an alias, `namedExpression` does:

```g4
namedExpression
    : expression (AS? (name=errorCapturingIdentifier | identifierList))?
    ;
```

I have changed that in 004bb692.

Reply:
Have have to come back to this:

It looks like Oracle and BigQuery allow for these aliases:

```sql
SELECT * FROM table UNPIVOT (
  val for var in (col1 AS alias1, col2 AS alias2)
)
```
and
```sql
SELECT * FROM table UNPIVOT (
  (val1, val2) for var in ((col1, col2) AS alias1, (col3, col4) AS alias2)
)
```

but still not
```sql
SELECT * FROM table UNPIVOT (
  (val1, val2) for var in ((col1 AS aliasA, col2 AS aliasB) AS alias1, (col3 AS aliasC, col4 AS aliasD) AS alias2)
)
```
(which was your original question).

https://docs.oracle.com/cd/B28359_01/server.111/b28286/statements_10002.htm#SQLRF55133
https://blogs.oracle.com/sql/post/how-to-convert-rows-to-columns-and-back-again-with-sql-aka-pivot-and-unpivot
https://hevodata.com/learn/bigquery-columns-to-rows/#u1

I'll add the former alias again.

Reply:
Added in 11cce9ef.